example/test/main.sh

```diff
@@ -16,100 +16,167 @@ fi
-# doing recovery next
-for jobnum in {1..5}; do
-  kill -9 %"${jobnum}"
-done
-
-microctl --state-dir "${test_dir}/c1" cluster list --local --format yaml |
-  yq '
-    sort_by(.name) |
-    .[0].role = "voter" |
-    .[1].role = "voter" |
-    .[2].role = "spare" |
-    .[3].role = "spare" |
-    .[4].role = "spare"' |
-  sed 's/:900/:800/' |
-  microctl --state-dir "${test_dir}/c1" cluster edit
-
-# While it is perfectly fine to load the recovery tarball on the member where it
-# was generated, the tests should make sure that both codepaths work, i.e. we
-# should make sure that recovery leaves the database ready to start with the
-# new configuration without needing to load the recovery tarball.
-mv "${test_dir}/c1/recovery_db.tar.gz" "${test_dir}/c2/"
-
-for member in c1 c2; do
-  state_dir="${test_dir}/${member}"
-  microd --state-dir "${state_dir}" "${cluster_flags[@]}" > /dev/null 2>&1 &
-done
-
-# microcluster takes a long time to update the member roles in the core_cluster_members table
-sleep 90
-
-microctl --state-dir "${test_dir}/c1" cluster list
-
-[[ $(microctl --state-dir "${test_dir}/c1" cluster list -f yaml | yq '.[] | select(.clustermemberlocal.name == "c1").role') == "voter" ]]
-[[ $(microctl --state-dir "${test_dir}/c1" cluster list -f yaml | yq '.[] | select(.clustermemberlocal.name == "c2").role') == "voter" ]]
-[[ $(microctl --state-dir "${test_dir}/c1" cluster list -f yaml | yq '.[] | select(.clustermemberlocal.name == "c3").role') == "spare" ]]
-[[ $(microctl --state-dir "${test_dir}/c1" cluster list -f yaml | yq '.[] | select(.clustermemberlocal.name == "c4").role') == "spare" ]]
-[[ $(microctl --state-dir "${test_dir}/c1" cluster list -f yaml | yq '.[] | select(.clustermemberlocal.name == "c5").role') == "spare" ]]
-
-echo "Tests passed"
-
-# Clean up
-if [ -n "${CLUSTER_INSPECT:-}" ]; then
-  echo "Pausing to inspect... press enter when done"
-  read -r
-fi
 
-for member in c1 c2; do
-  microctl --state-dir "${test_dir}/${member}" shutdown
-done
+  # The cluster doesn't always shut down right away; we've given it a chance
+  for job_pid in $(jobs -p); do
+    kill -9 "${job_pid}"
+  done
+}
+
+test_misc() {
+  new_systems 2
+
+    # Ensure two daemons cannot start in the same state dir
+  ! microd --state-dir "${test_dir}/c1" "${cluster_flags[@]}" || false
+
+  # Ensure only valid member names are used for bootstrap
+  ! microctl --state-dir "${test_dir}/c1" init "c/1" 127.0.0.1:9001 --bootstrap || false
+
+  microctl --state-dir "${test_dir}/c1" init "c1" 127.0.0.1:9001 --bootstrap
+
+  # Ensure only valid member names are used for join
+  token_node2=$(microctl --state-dir "${test_dir}/c1" tokens add "c/2")
+  ! microctl --state-dir "${test_dir}/c2" init "c/2" 127.0.0.1:9002 --token "${token_node2}" || false
+
+  shutdown_systems
+}
+
+test_tokens() {
+  new_systems 3
+  bootstrap_systems
+
+  microctl --state-dir "${test_dir}/c1" tokens add default_expiry
```

Comment:
Whilst reviewing this part I was wondering if we should add a comment to the microcluster `NewJoinToken()` func telling that by default there is no expiry?